core/src/main/scala/com/salesforce/op/stages/impl/tuning/DataCutter.scala

```diff
@@ -75,40 +78,91 @@ class DataCutter(uid: String = UID[DataCutter]) extends Splitter(uid = uid) with
 
   @transient private lazy val log = LoggerFactory.getLogger(this.getClass)
 
+  var cachedDataFrameForTesting: Option[DataFrame] = None
+
   /**
    * Function to set parameters before passing into the validation step
    * eg - do data balancing or dropping based on the labels
    *
    * @param data
    * @return Parameters set in examining data
    */
-  override def preValidationPrepare(data: Dataset[Row]): Option[SplitterSummary] = {
+  override def preValidationPrepare(data: DataFrame): PrevalidationVal = {
+    val labelColName = if (isSet(labelColumnName)) getLabelColumnName else data.columns(0)
+    val labelColIdx = data.columns.indexOf(labelColName)
+
+    if (!isSet(labelsToKeep)) {
+      val labelCounts = data.groupBy(labelColName).count().persist()
+      val res = estimate(labelCounts)
+      labelCounts.unpersist()
+      setLabels(res.labelsKept.distinct, res.labelsDropped.distinct, res.labelsDroppedTotal)
+    }
+
+    val labelMetaArr = getLabelsFromMetadata(data)
+    val labelSet = getLabelsToKeep.toSet
 
-    import data.sparkSession.implicits._
+    log.info(s"Dropping rows with columns not in $labelSet")
 
-    val labels = data.map(r => r.getDouble(0) -> 1L)
-    val labelCounts = labels.groupBy(labels.columns(0)).sum(labels.columns(1)).persist()
-    val (resKeep, resDrop) = estimate(labelCounts)
-    labelCounts.unpersist()
-    setLabels(resKeep, resDrop)
+    // Update metadata that spark.ml Classifier is using tp determine the number of classes
+    //
+    val na = NominalAttribute.defaultAttr.withName(labelColName)
+    val metadataNA = if (labelMetaArr.isEmpty) {
+      log.info("setting num vals " + labelSet.max.toInt + 1)
+      na.withNumValues(labelSet.max.toInt + 1)
+    } else {
+      val newLabelMetaArr = labelMetaArr
+        .zipWithIndex
+        .collect {
+          case (label: String, idx: Int) if labelSet.contains(idx.toDouble) => label
+        }
+
+      na.withValues(newLabelMetaArr)
+    }
 
-    summary = Option(DataCutterSummary(labelsKept = getLabelsToKeep, labelsDropped = getLabelsToDrop))
-    summary
+    // filter low cardinality labels out of the dataframe to reduce the volume and  to keep
+    // it in sync with the new metadata.
+    //
+    val dataPrep = data
+      .filter(r => labelSet.contains(r.getDouble(labelColIdx)))
+      .withColumn(labelColName, data(labelColName).as("_", metadataNA.toMetadata))
+
+    summary = Option(DataCutterSummary(
+      labelsKept = getLabelsToKeep,
+      labelsDropped = getLabelsToDrop,
+      labelsDroppedTotal = getLabelsDroppedTotal
+    ))
+
+    if (isSet(cacheValidatedDFForTesting) && $(cacheValidatedDFForTesting)) {
+      cachedDataFrameForTesting = Option(dataPrep)
+    }
```

Comment:
should we remove this before we merge?

Reply:
it would be good to keep the functionality to prevent regressions in the future way. I suppose it would be more elegant to have a TestDataCutter extending DataCutter with the sole purpose of intercepting the result dataframe.

Reply:
ok but lets make it private to OP if we just want it for testing